fix: Hide currency when amount is 0

daretobedifferent18 opened the pull request:
<!-- 
(Thanks for sending a pull request! Please make sure you click the link above to view the contribution guidelines, then fill out the blanks below.)
-->

<!-- Add the issue number that is fixed by this PR (In the form Fixes #123) -->

Fixes #5442 #5453 

#### Short description of what this resolves:
$ sign was shown even when amount was 0 so fixed this issue.

#### Changes proposed in this pull request:

- $ sign is not shown when amount is zero

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes <!-- use `ember test` to run all the tests -->
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

BEFORE 
![Screenshot from 2020-11-02 11-03-23](https://user-images.githubusercontent.com/71120226/97833294-0fa87780-1cfb-11eb-88bb-b8dd14b98663.png)

AFTER
![Screenshot from 2020-11-02 09-32-16](https://user-images.githubusercontent.com/71120226/97833310-1a630c80-1cfb-11eb-93d0-6ad576387418.png)
![Screenshot from 2020-11-02 09-32-33](https://user-images.githubusercontent.com/71120226/97833318-1fc05700-1cfb-11eb-9707-4001933413b9.png)
![Screenshot from 2020-11-02 09-32-56](https://user-images.githubusercontent.com/71120226/97833327-23ec7480-1cfb-11eb-8031-14557397616e.png)

daretobedifferent18 commented:
@mariobehling @iamareebjamal please see the changes made and let me know to do any more changes.

iamareebjamal commented on app/templates/components/public/ticket-list.hbs:
```diff
@@ -131,7 +131,7 @@
                     <td>{{currency-symbol this.eventCurrency}}</td> <td><Input @value={{this.overridenAmount}} placeholder="Total" /></td>
                   {{else}}
-                    <td>{{currency-symbol this.eventCurrency}} {{format-money this.orderAmount.total}}</td>
+                    <td>{{#if this.orderAmount}}{{currency-symbol this.eventCurrency}}{{/if}} {{format-money this.orderAmount.total}}</td>
```
This is not what was mentioned in the issue

daretobedifferent18 replied:
It says that $ sign should not be shown when the order amount is zero and if talking about country currency, if someone has added paid tickets in the event they also have to choose currency in the event so the currency they will be choosing will be reflected if amount is greater than zero, and i they have chosen only free tickets then no $ sign will be shown

daretobedifferent18 replied:
there was a bug which I didn't noticed 
Fixed now 
instead of ```this.orderAmount``` I have to use ```this.orderAmount.total```

daretobedifferent18 replied:
Please let me know if this is not the issue what I am thinking of.

daretobedifferent18 commented:
@iamareebjamal Please let me know if this is not the issue what I am thinking of.

iamareebjamal commented:
Please go ahead with this logic and solve other places in this PR as well

daretobedifferent18 commented:
Fixed Sales(amount)
![Screenshot from 2020-11-03 07-56-03](https://user-images.githubusercontent.com/71120226/97944557-00810280-1dab-11eb-8ab6-7f6257ad41fe.png)
---
Fixed sales and bottom column sales
![Screenshot from 2020-11-03 07-56-35](https://user-images.githubusercontent.com/71120226/97944559-01b22f80-1dab-11eb-91ef-c77fb4dc1c41.png)
---
Fixed amount for orders and attendees
![Screenshot from 2020-11-03 07-57-14](https://user-images.githubusercontent.com/71120226/97944560-02e35c80-1dab-11eb-9727-1b59bfcb38e9.png)
![Screenshot from 2020-11-03 07-57-41](https://user-images.githubusercontent.com/71120226/97944563-037bf300-1dab-11eb-8d98-fd0c2a51ee2e.png)
---
Fixed Discount (currency not showing)
![Screenshot from 2020-11-03 07-58-00](https://user-images.githubusercontent.com/71120226/97944565-04148980-1dab-11eb-9405-2dd1bb18a51c.png)
---
Fixed amount, subtotal (dashboard ticket page)
![Screenshot from 2020-11-03 07-58-45](https://user-images.githubusercontent.com/71120226/97944566-04ad2000-1dab-11eb-828c-40f9261bc76d.png)
---
Fixed amount, subtotal (public ticket page)
![Screenshot from 2020-11-03 08-00-03](https://user-images.githubusercontent.com/71120226/97944568-0545b680-1dab-11eb-8e4e-a5a679b89aa8.png)
---

iamareebjamal commented on app/templates/components/events/view/overview/event-tickets.hbs:
```diff
@@ -15,7 +15,7 @@
       <tr>
         <td>{{t 'Sales (amount)'}}</td>
-        <td>{{currency-symbol this.data.event.paymentCurrency}} {{format-money this.sales}}</td>
+        <td>{{#if this.sales}}{{currency-symbol this.data.event.paymentCurrency}}{{/if}} {{format-money this.sales}}</td>
```
Duplicate logic. There should be a single if statement in the entire PR

daretobedifferent18 replied:
You mean that like I have used if statement in different files, so there should be only one if or condition statement that should handle this?

daretobedifferent18 commented:
@iamareebjamal You mean that like I have used if statement in different files, so there should be only one if or condition statement that should handle this gloablly?

iamareebjamal commented:
Yes

daretobedifferent18 commented:
@iamareebjamal is this the correct way?
I modified currency-symbol file and passed one more arg to work accordingly
do let me know.

iamareebjamal commented on app/helpers/currency-symbol.js:
```diff
@@ -4,7 +4,10 @@ import { paymentCurrencies } from 'open-event-frontend/utils/dictionary/payment'
 export function currencySymbol(params) {
   const currency = find(paymentCurrencies, ['code', params[0]]);
-  return currency ? currency.symbol : params[0];
+  if(params[1])
```
This helper has nothing to do with amount and should not receive it

iamareebjamal commented on app/templates/components/events/view/overview/event-tickets.hbs:
```diff
@@ -15,7 +15,7 @@
       <tr>
         <td>{{t 'Sales (amount)'}}</td>
-        <td>{{currency-symbol this.data.event.paymentCurrency}} {{format-money this.sales}}</td>
+        <td>{{currency-symbol this.data.event.paymentCurrency this.sales}} {{format-money this.sales}}</td>
```
this.sales is repeating. It should only appear once

iamareebjamal commented on app/templates/components/orders/ticket-price.hbs:
```diff
@@ -1,5 +1,5 @@
     {{t "Free"}}
 {{else}}
-    {{currency-symbol @currency}} {{format-money @amount}}
+    {{currency-symbol @currency @amount}} {{format-money @amount}}
```
`@amount` is duplicated. It should only appear once

iamareebjamal commented:
No, now currency symbol is getting an unrelated field and amount is being duplicated. Think about what you do in JS when you want to remove duplicate code

daretobedifferent18 commented:
> No, now currency symbol is getting an unrelated field and amount is being duplicated. Think about what you do in JS when you want to remove duplicate code

@iamareebjamal ternary operator?

iamareebjamal commented:
Create a component

daretobedifferent18 commented:
@iamareebjamal I have added the component pls see to it

iamareebjamal commented on app/components/currency-amount.ts:
```diff
@@ -0,0 +1,4 @@
+import Component from '@glimmer/component';
+
+
+export default class CurrencyAmount extends Component {}
```
Delete this file, no need of it

iamareebjamal commented on app/templates/components/events/view/overview/event-tickets.hbs:
```diff
@@ -15,7 +15,7 @@
       <tr>
         <td>{{t 'Sales (amount)'}}</td>
-        <td>{{currency-symbol this.data.event.paymentCurrency this.sales}} {{format-money this.sales}}</td>
+        <td><CurrencyAmount @currency={{currency-symbol this.data.event.paymentCurrency}} @amount={{format-money this.sales}}/></td>
```
What's the point of component if you are doing calculations here. currency-symbol should not be called here. It should be called inside the component

daretobedifferent18 replied:
ok but for amount I have to do that it here only bcz for diff it is diff

iamareebjamal commented on app/templates/components/currency-amount.hbs:
```diff
@@ -0,0 +1 @@
+{{#if (gt @amount 0)}} {{@currency}} {{/if}}{{@amount}}
```
There is no space between currency and amount

daretobedifferent18 replied:
its there

iamareebjamal commented:
Half files are using component, half are not. `currency-symbol` helper is duplicated in all component calls

daretobedifferent18 commented:
> Half files are using component, half are not. `currency-symbol` helper is duplicated in all component calls

in some places I have to take care of only currency symbol so for that I think I should not call the comp. and just do it there only?

iamareebjamal commented:
Let's say you have a function which returns x^2 if x is even and 1/x if x is odd. Choose what the function and function call should look like:

1. x if y else z | f(x^2, x % 2, 1/x)
2. x if y % 2 else z | f(x^2, x, 1/x)
3. x^2 if x % 2 else 1/x | f(x)

daretobedifferent18 commented:
> Let's say you have a function which returns x^2 if x is odd and 1/x if x is even. Choose what the function and function call should look like:
> 
>     1. x if y else z | f(x^2, x % 2, 1/x)
> 
>     2. x if y % 2 else z | f(x^2, x, 1/x)
> 
>     3. x^2 if x % 2 else 1/x | f(x)

3

daretobedifferent18 commented:
@iamareebjamal 

```handlebars
{{this.model.eventDetail.paymentCurrency}}
{{sub ticket.price ticket.discount}}
{{mult ticket.orderStatistics.tickets.completed ticket.price}}
```
so now there are diff cases where sometimes there is only arg directly passing or sometimes evaluating before and then passing as a single arg doing mul or sub 
so should I handle this before calculation also in comp.??

daretobedifferent18 commented:
I have some changes in the component

iamareebjamal commented on app/templates/events/view/tickets/index.hbs:
```diff
@@ -144,15 +144,15 @@
             <td class="right aligned">{{ticket.orderStatistics.tickets.completed}}</td>
           {{/unless}}
-          <td class="right aligned">{{currency-symbol this.model.eventDetail.paymentCurrency}} {{mult ticket.orderStatistics.tickets.completed ticket.price}}</td>
+          <td class="right aligned"><CurrencyAmount @currency={{this.model.eventDetail.paymentCurrency}} @format={{false}} @amount={{mult ticket.orderStatistics.tickets.completed ticket.price}}/></td>
```
The amount here should still be formatted. The fact that it wasn't was a bug

iamareebjamal commented on app/templates/components/currency-amount.hbs:
```diff
@@ -0,0 +1,9 @@
+    {{currency-symbol @currency}}
+{{/if}}
+
+{{#if (eq @format false)}}
```
There is no case amount shouldn't be formatted

daretobedifferent18 replied:
there is a place in ticket overview page where amount was not being formated so acc to that but if that was a bug? that I will fix this

iamareebjamal commented:
> so should I handle this before calculation also in comp.??

No

iamareebjamal commented on tests/integration/components/currency-amount-test.js:
```diff
@@ -0,0 +1,18 @@
+
+    await render(hbs`{{currency-amount}}`);
+
+    assert.equal(this.element.textContent.trim(), '');
```
Add tests for different cases

daretobedifferent18 replied:
ok and rest changes in file is ok?

iamareebjamal commented:
Ensure there is a space between symbol and amount

daretobedifferent18 commented:
> Ensure there is a space between symbol and amount
See it there and also same in various other places
![Screenshot from 2020-11-05 14-42-52](https://user-images.githubusercontent.com/71120226/98221079-370a7900-1f75-11eb-8996-7d261ddab436.png)
Is this right?

daretobedifferent18 commented:
![Screenshot from 2020-11-05 15-12-03](https://user-images.githubusercontent.com/71120226/98225241-80a99280-1f7a-11eb-938b-5494f8708e1e.png)

![Screenshot from 2020-11-05 15-26-05](https://user-images.githubusercontent.com/71120226/98225822-3c6ac200-1f7b-11eb-9f31-4fed8f7ece16.png)
